### DISCUSSION TITLE
Need support in writing a preview plugin that supports decode and preview JWT files.

### DISCUSSION THREAD

**Discussion by smangels** [sxyazi/yazi#2109] [1. Q&A]
### What system are you running Yazi on?

Linux X11

### What terminal are you running Yazi in?

kitty 0.38.1 and ghostty 1.0.0 build 8469

### `yazi --debug` output

```Shell
Yazi
    Version: 0.4.2 (Homebrew 2024-12-20)
    Debug  : false
    Triple : aarch64-apple-darwin (macos-aarch64)
    Rustc  : 1.83.0 (90b35a62 2024-11-26)

Ya
    Version: 0.4.2 (Homebrew 2024-12-20)

Emulator
    Brand.from_env      : Some(Ghostty)
    Emulator.detect     : Emulator { kind: Left(Ghostty), light: false, cell_size: Some((20, 41)) }
    Emulator.detect_full: Ok(Emulator { kind: Left(Ghostty), light: false, cell_size: Some((20, 41)) })

Adapter
    Adapter.matches: Kgp

Desktop
    XDG_SESSION_TYPE           : None
    WAYLAND_DISPLAY            : None
    DISPLAY                    : None
    SWAYSOCK                   : None
    HYPRLAND_INSTANCE_SIGNATURE: None
    WAYFIRE_SOCKET             : None

SSH
    shared.in_ssh_connection: false

WSL
    WSL: false

NVIM
    NVIM          : false
    Neovim version: 0.10.3

Variables
    SHELL           : Some("/bin/zsh")
    EDITOR          : Some("nvim")
    VISUAL          : None
    YAZI_FILE_ONE   : None
    YAZI_CONFIG_HOME: None
    YAZI_ZOXIDE_OPTS: None
    FZF_DEFAULT_OPTS: Some("\n    --color=fg:#d0d0d0,fg+:#d0d0d0,bg:-1,bg+:#262626\n    --color=hl:#5f87af,hl+:#5fd7ff,info:#afaf87,marker:#87ff00\n    --color=prompt:#d7005f,spinner:#af5fff,pointer:#af5fff,header:#87afaf\n    --color=border:#7aa2f7,label:#aeaeae,query:#d9d9d9\n    --height 60% --extended --border")

Text Opener
    default     : Some(Opener { run: "${EDITOR:-vi} \"$@\"", block: true, orphan: false, desc: "$EDITOR", for_: None, spread: true })
    block-create: Some(Opener { run: "${EDITOR:-vi} \"$@\"", block: true, orphan: false, desc: "$EDITOR", for_: None, spread: true })
    block-rename: Some(Opener { run: "${EDITOR:-vi} \"$@\"", block: true, orphan: false, desc: "$EDITOR", for_: None, spread: true })

Multiplexers
    TMUX               : 0
    tmux version       : tmux 3.5a
    tmux build flags   : enable-sixel=Unknown
    ZELLIJ_SESSION_NAME: None
    Zellij version     : 0.41.2

Dependencies
    file          : 5.41
    ueberzugpp    : No such file or directory (os error 2)
    ffmpeg/ffprobe: No such file or directory (os error 2) / No such file or directory (os error 2)
    pdftoppm      : No such file or directory (os error 2)
    magick        : No such file or directory (os error 2)
    fzf           : 0.57.0
    fd/fdfind     : 10.2.0 / No such file or directory (os error 2)
    rg            : 14.1.1
    chafa         : No such file or directory (os error 2)
    zoxide        : 0.9.6
    7z/7zz        : 17.05 / No such file or directory (os error 2)
    jq            : 1.7.1

Clipboard
    wl-copy/paste: No such file or directory (os error 2) / No such file or directory (os error 2)
    xclip        : No such file or directory (os error 2)
    xsel         : No such file or directory (os error 2)
```

### Describe the question

Hi,
I'm reaching out for people willing to support me in writing a native previewer plugin for JWT files. Nats server and account configuration is stored in such a format and when navigating through hundreds of JWT files in a file structure for a cluster, it would be nice to have native preview support in Yazi (with code highlighting of the resulting JSON content inside the payload).
Many thanks in advance.

### Checklist

- [X] I have searched the existing discussions/issues

**Comment by AnirudhG07** [comment]
Hi! @smangels Mostly any yazi "previewers" are based on "cli" tools that show (only) the preview for it in the terminal. Many JWT CLI's I have seen also allow editing and other functionalities, than just a good preview.
If you know of something of a JWT previewer, I can make it for you.
Or
If you are developing a new tool itself, I would like to help out(if possible).

**Comment by smangels** [comment]
Hi,
I've managed to get it up an running decoding the JWT through a CLI tool and are happy to share the code with you. It would be even better if we could get it decode natively in Lua. But I've struggled to get a development environment up and running. Did you manage Anirudh?
/Sebastian 


BR,

Re:

**Reply by AnirudhG07** [reply]
We need not convert it to a Lua code, if the CLI can be installed, and can just preview nicely, then it should be enough in that language itself.
Can you share the repo? And a screenshot of the preview of how it looks?

**Reply by sxyazi** [reply]
> It would be even better if we could get it decode natively in Lua

Yazi has a `ya.json_decode()` function, here's an example:

https://github.com/sxyazi/yazi/blob/63eb82aa72ed43e84571418aaa56d90383358c28/yazi-plugin/preset/plugins/video.lua#L115-L121

**Comment by smangels** [comment]
The repository can be found at https://github.com/smangels/jwt.yazi and it contains a test.jwt file. The issue is that we are using our own JWT decoder binary and my gut feeling tells me that it would be much cleaner to simply implement the decoder feature in plain Lua instead of relying on an external tool - shouldn't be too complicated.

https://gist.github.com/inaiat/02bf5d11732d8e4d7b7546399a3a49af

What do you think?

Re:

**Reply by AnirudhG07** [reply]
Yeah it would be great, you can add this in your init.lua and call the function.
I feel its better if the plugin doesnt have external dependency if it can be implemented within `init.lua` itself.

**Comment by smangels** [comment]
What I meant was rather decoding JWT natively in Lua. Instead of having to rely on external CLI tool.

BR,

Re:

**Comment by smangels** [comment]
Yeah I agree. Do you know how to make the Lua environment aware of that there the Yazi Lua library somewhere in users file structure? I'm using Neovim.



BR,

Re:

**Comment by AnirudhG07** [comment]
Created PR [#1](https://github.com/smangels/jwt.yazi/pull/1) in `jwt.yazi`. Its not perf since I am SOMEHOW not able to read `test.jwt` file( seriously I donno why its not working, must be something stupid i am messing up).

Rest I think I have mentioned in the PR. Cool!

**Comment by smangels** [comment]
I'll have a look into it. Could be the process of splitting up header, body and signature? Or maybe the fact that padding is applied when length is uneven?


BR,

Re:
